File: data/src/main/java/data/sections/users/GithubUsersApi.java
/*
 * Copyright 2016 Victor Albertos
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package data.sections.users;

import java.util.List;

import io.reactivex.Single;
import retrofit2.Response;
import retrofit2.http.GET;
import retrofit2.http.Headers;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface GithubUsersApi {
    String HEADER_API_VERSION = "Accept: application/vnd.github.v3+json";

    @Headers({HEADER_API_VERSION})
    @GET("/users/{username}")
    Single<Response<User>> getUserByName(@Path("username") String username);

    @Headers({HEADER_API_VERSION})
    @GET("/users")
    Single<Response<List<User>>> getUsers(@Query("since") int lastIdQueried, @Query("per_page") int perPage);
}

File: app/src/main/kotlin/io.victoralbertos.kdirtyandroid/network/api/GithubUsersApi.kt
package io.victoralbertos.kdirtyandroid.network.api

import io.victoralbertos.kdirtyandroid.network.dtos.UserDTO
import retrofit2.http.GET
import retrofit2.http.Headers
import retrofit2.http.Query

interface GithubUsersApi {
    @Headers(HEADER_API_VERSION)
    @GET("/users")
    suspend fun getUsers(
        @Query("since") lastIdQueried: Int,
        @Query("per_page") perPage: Int
    ): List<UserDTO>

    companion object {
        private const val HEADER_API_VERSION = "Accept: application/vnd.github.v3+json"
    }
}
